swift/extractor/Extractor.cpp

```diff
@@ -0,0 +1,101 @@
+#include "Extractor.h"
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <memory>
+#include <unistd.h>
+
+#include <swift/AST/SourceFile.h>
+#include <llvm/ADT/SmallString.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Path.h>
+
+using namespace codeql;
+
+Extractor::Extractor(const Configuration& config, swift::CompilerInstance& instance)
+    : config{config}, compiler{instance} {}
+
+void Extractor::extract() {
+  if (compiler.getPrimarySourceFiles().empty()) {
```

Comment:
What's the reason for this `if`? In other words, why would be primary source files be missing?

Reply:
I guess it's worth adding a comment there.
In general, there are many ways the frontend can be called:
```
swift-frontend -c main.swift # compiles main.swift
swift-frontend -primary-file a.swift b.swift c.swift # compiles a.swift only
swift-frontend <some other command to build a swift module from precompiled files, no source files involved>
```

Reply:
Are all these paths even possible in the context of the extractor?

Reply:
Yes, they are. The tracer invokes the extractor without knowing exactly what the swift-frontend does. It's our job later to decide what to do with all the arguments/invocation.